src/relay/transforms/to_a_normal_form.cc

```diff
@@ -81,171 +70,187 @@ std::unordered_map<DependencyGraph::Node*, Scope> CalcScope(const DependencyGrap
+                                  NodeScopeMap* node_scope, ExprSet* lifted) {
+  Fill fi(dg, node_scope, lifted);
+  auto var = fi.VisitExpr(e);
+  return fi.GetScope(e)->ll->Get(var);
+}
 
-  Scope GetScope(const Expr& e) { return node_scope_->at(dg_.expr_node.at(e)); }
+Scope Fill::GetScope(const Expr& e) { return node_scope_->at(dg_.expr_node.at(e)); }
 
-  Scope GetSubScope(const Expr& e, size_t i) {
-    DependencyGraph::Node* n = dg_.expr_node.at(e);
-    auto h = n->children.head;
-    while (i != 0) {
-      CHECK(h);
-      --i;
-      h = h->next;
-    }
+Scope Fill::GetSubScope(const Expr& e, size_t i) {
+  DependencyGraph::Node* n = dg_.expr_node.at(e);
+  auto h = n->children.head;
+  while (i != 0) {
     CHECK(h);
-    return node_scope_->at(h->value);
+    --i;
+    h = h->next;
   }
+  CHECK(h);
+  return node_scope_->at(h->value);
+}
 
-  Expr VisitExpr(const Expr& e, const Var& v) final {
-    if (memo.count(e) == 0) {
-      memo.insert({e, ExprFunctor<Expr(const Expr&, const Var&)>::VisitExpr(e, v)});
-    } else if (v.defined()) {
-      GetScope(e)->ll->Push(v, memo.at(e));
-    }
-    auto ret = memo.at(e);
-    CHECK(IsAtomic(ret));
-    return ret;
+Expr Fill::VisitExpr(const Expr& e, const Var& v) {
+  if (memo.count(e) == 0) {
+    memo.insert({e, ExprFunctor<Expr(const Expr&, const Var&)>::VisitExpr(e, v)});
+  } else if (v.defined()) {
+    GetScope(e)->ll->Push(v, memo.at(e));
   }
+  auto ret = memo.at(e);
+  // if no include_set is specified, every expression should be atomic.
+  if (include_set_ == nullptr) CHECK(IsAtomic(ret));
+  return ret;
+}
 
-  Expr VisitExpr(const Expr& e) { return this->VisitExpr(e, Var()); }
+Expr Fill::VisitExpr(const Expr& e) { return this->VisitExpr(e, Var()); }
 
-  Expr Atomic(const Expr& e, const Var& v) { return v.defined() ? GetScope(e)->ll->Push(v, e) : e; }
+Expr Fill::Atomic(const Expr& e, const Var& v) {
+  return v.defined() ? GetScope(e)->ll->Push(v, e) : e;
+}
 
-  Expr Compound(const Expr& orig, const Expr& now, const Var& v) {
-    Var var = v.defined() ? v : Var(String("x"), Type());
+// Bind expression `now` to var `v` if the original expression is in the include set, or if
+// v is already defined (e.g. coming from a Let expression). Otherwise return `now` directly
+Expr Fill::Compound(const Expr& orig, const Expr& now, const Var& v) {
+  Var var = v.defined() ? v : Var(String("x"), Type());
+  bool not_included = include_set_ && include_set_->find(orig) == include_set_->end();
+  if (!v.defined() && not_included) {
+    return now;
+  } else {
     return GetScope(orig)->ll->Push(var, now);
   }
+}
 
-  Expr VisitExpr_(const CallNode* c, const Var& v) final {
-    Expr e = GetRef<Expr>(c);
-    std::vector<Expr> args;
-    for (const auto& a : c->args) {
-      args.push_back(VisitExpr(a));
-    }
-    return Compound(e, Call(VisitExpr(c->op), args, c->attrs, c->type_args), v);
+Expr Fill::VisitExpr_(const CallNode* c, const Var& v) {
```

Comment:
Why remove `final` from all of these functions?

Reply:
`final` is still used in the header at declaration https://en.cppreference.com/w/cpp/language/final#Syntax